src/sage/rings/function_field/drinfeld_modules/drinfeld_module.py

```diff
@@ -1532,9 +1696,96 @@ def velu(self, isog):
-        if not char_deg.divides(isog.valuation()) \
-                or rem != 0:
-            raise e
-        else:
+        if rem.is_zero() and quo[0] == self.gen()[0]:
             return self.category().object(quo)
+        else:
+            raise e
+
+    def hom(self, x, codomain=None):
+        r"""
+        Return the homomorphism defined by ``x`` having this Drinfeld
+        module as domain.
+
+        We recall that a homomorphism `f : \phi \to \psi` between
+        two Drinfeld modules is defined by a Ore polynomial `u`,
+        which is subject to the relation `phi_T u = u \psi_T`.
+
+        INPUT:
+
+        - ``x`` -- an element of the ring of functions, or a
+          Ore polynomial
+
+        - ``codomain`` -- a Drinfeld module or ``None`` (default:
+          ``None``)
+
+        EXAMPLES::
+
+            sage: Fq = GF(5)
+            sage: A.<T> = Fq[]
+            sage: K.<z> = Fq.extension(3)
+            sage: phi = DrinfeldModule(A, [z, 0, 1, z])
+            sage: phi
+            Drinfeld module defined by T |--> z*t^3 + t^2 + z
+
+        An important class of endomorphisms of a Drinfeld module
+        `\phi` is given by scalar multiplications, that are endomorphisms
+        corresponding to the Ore polynomials `\phi_a` with `a` in the function
+        ring `A`. We construct them as follows::
+
+            sage: phi.hom(T)
+            Endomorphism of Drinfeld module defined by T |--> z*t^3 + t^2 + z
+              Defn: z*t^3 + t^2 + z
+
+            sage: phi.hom(T^2 + 1)
+            Endomorphism of Drinfeld module defined by T |--> z*t^3 + t^2 + z
+              Defn: z^2*t^6 + (3*z^2 + z + 1)*t^5 + t^4 + 2*z^2*t^3 + (3*z^2 + z + 1)*t^2 + z^2 + 1
+
+        We can also define a morphism by passing in the Ore polynomial
+        defining it.
+        For example, below, we construct the Frobenius endomorphism
+        of `\phi`::
+
+            sage: t = phi.ore_variable()
+            sage: phi.hom(t^3)
+            Endomorphism of Drinfeld module defined by T |--> z*t^3 + t^2 + z
+              Defn: t^3
+
+        If the input Ore polynomial defines a morphism to another
+        Drinfeld module, the latter is determined automatically::
+
+            sage: phi.hom(t + 1)
+            Drinfeld Module morphism:
+              From: Drinfeld module defined by T |--> z*t^3 + t^2 + z
+              To:   Drinfeld module defined by T |--> (2*z^2 + 4*z + 4)*t^3 + (3*z^2 + 2*z + 2)*t^2 + (2*z^2 + 3*z + 4)*t + z
+              Defn: t + 1
+
+        TESTS::
+
+            sage: phi.hom(t)
+            Traceback (most recent call last):
+            ...
+            ValueError: the input does not define an isogeny
+
+        ::
+
+            sage: phi.hom(T + z)
+            Traceback (most recent call last):
+            ...
+            ValueError: the input does not define an isogeny
+
+        """
+        if self.function_ring().has_coerce_map_from(x.parent()):
+            return self.Hom(self)(x)
```

Comment:
> Beside, should `DrinfedModule._End_` be created?

No, it's not useful. The construction `End(X)` calls `Hom(X,X)` which itself calls `X._Hom_(X)`.

Reply:
> This line of code made me realize that the current documentation was not up to date. In particular, it is not specified in the docstrings of `DrinfeldModuleHomset._element_constructor_`, `DrinfeldModuleMorphism.__init__` and `DrinfeldModuleMorphism.__classcall_private__` that the input may be an element in the function ring (or something that coerces to it). Some parameters must also have their names changed. This problem must be addressed and I can volunteer to proceed, if you want.

IMNO, the current code+doctest is fine. What would you like to change?

Reply:
> What would you like to change?

If I take `DrinfeldModuleMorphism.__classcall_private__`, the argument `x` may be:
- an Ore polynomial;
- a `DrinfeldModuleMorphism`;
- an element in the function ring.

This last point is not specified in the documentation:

```
        - ``x`` -- the Ore polynomial defining the morphism or a
          DrinfeldModuleMorphism
```

nor showcased in the examples. Code is fine. But I believe the documentation should be upgrade to explain that the input may be an element in the function ring! No?

I realized that `DrinfeldModuleMorphism.__init__` is fine.

Similarly, the documentation of `DrinfeldModuleHomset._element_constructor_` specifies that the input is an Ore polynomial, but it can also be an element in the function ring.

Reply:
Done

Reply:
Thank you very much for taking this time! Next time I can do it, if you want.